src/metadata/selectors.js

```diff
@@ -1,229 +1,214 @@
-    )
-}
+export const getGroupedDataElementsByCatComboInOrder = createSelector(
+    (_, dataElements) => dataElements,
+    getCategoryCombos,
+    (dataElements, categoryCombos) => {
+        const grouped = []
+
+        dataElements.forEach((de, i) => {
+            const previousDe = dataElements[i - 1]
+            const lastGroup = grouped[grouped.length - 1]
+            const matchingId =
+                previousDe?.categoryCombo.id === de.categoryCombo.id
+
+            if (!lastGroup || !matchingId) {
+                grouped.push({
+                    dataElements: [de],
+                    categoryCombo: categoryCombos[de.categoryCombo.id],
+                })
+            } else {
+                lastGroup.dataElements.push(de)
+            }
+        })
+
+        return grouped
+    }
+)
 
 /**
- * Tries to find the categoryOptionCombo with the given categoryOptions within
- * a category combination.
- * @param {} metadata
- * @param {*} catComboId
+ * Returns an array of objects with dataElements and their associated categoryCombos. Unlike
+ * getGroupedDataElementsByCatComboInOrder, this selector will group all dataElements with the
+ * same categoryComboId together.
+ * @param {*} metadata
+ * @param {*} dataElements
+ */
+export const getGroupedDataElementsByCatCombo = createSelector(
+    (_, dataElements) => dataElements,
+    getCategoryCombos,
+    (dataElements, categoryCombos) => {
+        // Group dataElements by their categoryCombo id
+        const grouped = groupBy(dataElements, (de) => de.categoryCombo.id)
+
+        // Map to an array and include the associated categoryCombo
+        return Object.entries(grouped).map(
+            ([categoryComboId, dataElements]) => ({
+                dataElements,
+                categoryCombo: categoryCombos[categoryComboId],
+            })
+        )
+    }
+)
+
+/**
+ * Tries to find the categoryOptionCombo with the given categoryOptions within a category
+ * combination.
+ * @param {*} metadata
+ * @param {*} categoryComboId
  * @param {*} categoryOptionIds
- * @returns
  */
-export const getCoCByCategoryOptions = (
-    metadata,
-    catComboId,
-    categoryOptionIds
-) => {
-    const cocs = getCategoryOptionCombosByCategoryComboId(metadata, catComboId)
-    const sorted = [...categoryOptionIds].sort() //sort to check for equality
-
-    for (const coc of cocs) {
-        const sortedCatOptions = [...coc.categoryOptions].sort()
-        if (
-            sorted.length === sortedCatOptions.length &&
-            sortedCatOptions.every((id, ind) => id === sorted[ind])
-        ) {
-            return coc
+export const getCoCByCategoryOptions = createCachedSelector(
+    getCategoryOptionCombosByCategoryComboId,
+    (_, categoryComboId) => categoryComboId,
+    (_, __, categoryOptionIds) => categoryOptionIds,
+    (categoryOptionCombos, categoryComboId, categoryOptionIds) => {
+        // Sort since we only care about the contents being identical, not the order
```

Comment:
Instead of sorting arrays, we could use `Set`s? (i.e. `const targetIds = new Set(categoryOptionIds)`)

Reply:
Do you know if that's safe to use with our browser support? I've not looked into that recently

Reply:
Even IE11 supports it according to [this MDN table](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set#browser_compatibility) so should be alright imo

Reply:
Ah ok nice. Yeah should be fine then

Reply:
After taking a look at the implementation that used Set I went in a slightly different direction. Because we don't need the sorting at all actually.

Since we don't care about the order, it's sufficient to check if the arrays have the same length and that all ids in targetIds are included in currentIds. That yielded a solution that I think is simple enough and quite readable. I can imagine the Set might be more performant, but I felt readability trumped performance here.